leader/leader.go

```diff
@@ -105,64 +108,71 @@ func (l *Client) AddWatchCallback(key string, retry time.Duration, fn CallbackFn
-	if err != nil {
-		return nil, nil, trace.BadParameter("%v unexpected error: %v", ctx.Value("prefix"), err)
-	} else if resp == nil {
-		log.Debugf("%v client is closing, return", ctx.Value("prefix"))
-		return nil, nil, nil
+func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI, key string, retry time.Duration, valuesC chan string) client.Watcher {
+	logger := log.WithField("key", key)
+	logger.Info("Recreating watch at the latest index.")
+	resp := l.getFirstValue(key, retry)
+	if resp == nil {
+		logger.Info("Client is closing.")
+		return nil
 	}
-	log.Debugf("%v got current value '%v' for key '%v'", ctx.Value("prefix"), resp.Node.Value, key)
+	// After reestablishing the watch, always send the value we got to the client.
+	if resp.Node != nil {
+		logger.WithFields(logrus.Fields{
+			"value": resp.Node.Value,
+			"index": resp.Index,
+		}).Info("Got current value.")
+		select {
+		case valuesC <- resp.Node.Value:
+		case <-l.closeC:
+			logger.Info("Watcher has been closed.")
+			return nil
+		}
+	}
+	// The watcher that will be receiving events after the value we got above.
 	watcher := api.Watcher(key, &client.WatcherOptions{
 		// Response.Index corresponds to X-Etcd-Index response header field
 		// and is the recommended starting point after a history miss of over
 		// 1000 events
 		AfterIndex: resp.Index,
 	})
-	return watcher, resp, nil
+	return watcher
 }
 
 // AddWatch starts watching the key for changes and sending them
-// to the valuesC, the watch is stopped
+// to the valuesC until the client is stopped.
 func (l *Client) AddWatch(key string, retry time.Duration, valuesC chan string) {
-	prefix := fmt.Sprintf("AddWatch(key=%v)", key)
 	api := client.NewKeysAPI(l.client)
 
+	logger := log.WithField("key", key)
+	logger.WithFields(logrus.Fields{
+		"peers": l.ETCD.Endpoints,
+		"retry": retry,
+	}).Info("Setting up watch.")
+
 	go func() {
 		var watcher client.Watcher
 		var resp *client.Response
 		var err error
 
-		ctx, closer := context.WithCancel(context.WithValue(context.Background(), "prefix", prefix))
+		ctx, closer := context.WithCancel(context.Background())
 		go func() {
 			<-l.closeC
 			closer()
 		}()
 
 		backoff := NewUnlimitedExponentialBackOff()
 		ticker := ebackoff.NewTicker(backoff)
-		var steps int
-
-		watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-		if err != nil {
-			return
-		}
-
-		// make sure we always send the first actual value
-		if resp != nil && resp.Node != nil {
-			select {
-			case valuesC <- resp.Node.Value:
-			case <-l.closeC:
-				return
-			}
-		}
 
+		var steps int
 		var sentAnything bool
-		for {
 
+		for {
+			// We're either establishing a new watch or it has been reset
+			// due to an error/expiration so create a new one at the most
+			// recent index.
 			if watcher == nil {
-				watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
+				watcher = l.getWatchAtLatestIndex(ctx, api, key, retry, valuesC)
```

Comment:
This is also confusing in that it might be nil which is not super clear why.
Why not return both watcher and error as before so it's clear when it fails (which can be due to close event in this case)?
Also, if the `watcher` was not created due to close event (or any other error really) and `err` is also nil - the `resp` might be nil as well and will generate nil pointer.

Reply:
Like I mentioned above, this function never actually returns an error currently. It only returns a nil watcher if the client closes. I think that returning an error here will only complicate things for the caller b/c it will have to additionally process the error without much benefit.

I think what I missed here, is that if we still got a nil watcher here, we should just return because we know the client has closed. Then there won't be a nil deref.